engine/src/main/java/org/camunda/bpm/engine/delegate/BpmnError.java

```diff
@@ -43,16 +43,38 @@
   private String errorMessage;
 
   public BpmnError(String errorCode) {
-    super("");
+    super(exceptionMessage(errorCode, null));
     setErrorCode(errorCode);
   }
           
   public BpmnError(String errorCode, String message) {
-    super(message + " (errorCode='" + errorCode + "')");
+    super(exceptionMessage(errorCode, message));
     setErrorCode(errorCode);
     setMessage(message);
   }
 
+  public BpmnError(String errorCode, String message, Throwable cause) {
+    super(exceptionMessage(errorCode, message), cause);
+    setErrorCode(errorCode);
+    setMessage(exceptionMessage(errorCode, message) + (cause != null ? "; details: " + cause.getMessage() : ""));
```

Comment:
I just noticed that `BpmnError#getMessage` overrides `Throwable#getMessage`. In all other constructors where a `message` parameter is provided, the exception message is just set to that value. I think we should stick to that behavior for all constructors, i.e. not concatenate error message, code and cause here. That way, calling code can decide if it wants to concatenate the cause message to the error message or if its sufficient if the cause is referenves as the causing exception. What do you think?

Reply:
Until now, it was not possible to provide a cause. Hence it was not possible to take over (or concatenate) the message from the cause. One could decide either way. Some will think like you wrote, some others (like me :-) would expect that the cause message is integrated into the explicitly provided message (if I only want the message I explicitly provided, I won't specify the cause).

The problem stems from the fact that the error message overrides the "getMessage" method. This results in a somewhat ugly code. If there was a separate method (e.g. "getErrorMessage") then the getMessage would work like it does in all other throwables. In this class, its semantics has been changed. E.g.: Why does it include the error code if no message has been specified?

If the message from the cause is not integrated into the BpmnError message, it's effectively lost since the cause is not available when the BpmnError is catched (i.e. in the error event), only code and message are available. Why then specify the cause at all? Just to make sonar happy? Too little in my view.

You decide.

Reply:
> Until now, it was not possible to provide a cause. Hence it was not possible to take over (or concatenate) the message from the cause. One could decide either way. Some will think like you wrote, some others (like me :-) would expect that the cause message is integrated into the explicitly provided message (if I only want the message I explicitly provided, I won't specify the cause).

I agree. I would prefer if we could go with the more flexible option that allows the user to define the message.

> The problem stems from the fact that the error message overrides the "getMessage" method. This results in a somewhat ugly code. If there was a separate method (e.g. "getErrorMessage") then the getMessage would work like it does in all other throwables. 

Agree that a separater getter for the error message would be better. 

> In this class, its semantics has been changed. E.g.: Why does it include the error code if no message has been specified?

I don't think it does. I assume you are referring to the following line: https://github.com/camunda/camunda-bpm-platform/blob/7.12.0/engine/src/main/java/org/camunda/bpm/engine/delegate/BpmnError.java#L51. The constructor includes the error code in a String that it hands to the super-constructor, but since `#getMessage` is overridden and always returns the value of the field `errorMessage`, that does not influence any behavior. Maybe I am misunderstanding your point.

> If the message from the cause is not integrated into the BpmnError message, it's effectively lost since the cause is not available when the BpmnError is catched (i.e. in the error event), only code and message are available. Why then specify the cause at all? Just to make sonar happy? Too little in my view.

I understand your point, yet I would argue that this is okay. You still have the choice as the user to include the cause exception message with any formatting you like. In addition, the cause may be useful for features we may build in the future (e.g. debug logging of BPMN errors; making the error object accessible in the execution context of the catching boundary event).

Reply:
You can stripp off the cause message from the BpmnError message, I won't be offended. Change it as you like. Your points are also valid.